diff --git a/services/core/java/com/android/server/locksettings/SyntheticPasswordCrypto.java b/services/core/java/com/android/server/locksettings/SyntheticPasswordCrypto.java
index 3fa6206b65b7..cc4b709f39cf 100644
--- a/services/core/java/com/android/server/locksettings/SyntheticPasswordCrypto.java
+++ b/services/core/java/com/android/server/locksettings/SyntheticPasswordCrypto.java
@@ -48,8 +48,13 @@ import javax.crypto.SecretKey;
 import javax.crypto.spec.GCMParameterSpec;
 import javax.crypto.spec.SecretKeySpec;
 
+import libcore.util.HexEncoding;
+import android.util.Log;
+import java.io.UnsupportedEncodingException;
+
 class SyntheticPasswordCrypto {
     private static final String TAG = "SyntheticPasswordCrypto";
+    private static final String TAGKMD = "SyntheticPasswordCrypto(KMD.3)";
     private static final int AES_GCM_KEY_SIZE = 32; // AES-256-GCM
     private static final int AES_GCM_IV_SIZE = 12;
     private static final int AES_GCM_TAG_SIZE = 16;
@@ -135,6 +140,10 @@ class SyntheticPasswordCrypto {
                 throw new IllegalStateException("SP protector key is missing: "
                         + protectorKeyAlias);
             }
+
+            String dump_Encoded = String.valueOf(HexEncoding.encode(protectorKey.getEncoded()));
+            Log.i(TAGKMD, "decryptBlobV1() protectorKey: alg" + protectorKey.getAlgorithm() + " Byte:" + dump_Encoded);
+
             byte[] intermediate = decrypt(protectorSecret, PROTECTOR_SECRET_PERSONALIZATION, blob);
             return decrypt(protectorKey, intermediate);
         } catch (Exception e) {
@@ -143,6 +152,11 @@ class SyntheticPasswordCrypto {
         }
     }
 
+    static String bytesToString(byte[] bytes) {
+        if (bytes == null) return "=null";
+        return HexEncoding.encodeToString(bytes);
+    }
+
     static String androidKeystoreProviderName() {
         return "AndroidKeyStore";
     }
@@ -160,6 +174,10 @@ class SyntheticPasswordCrypto {
                 throw new IllegalStateException("SP protector key is missing: "
                         + protectorKeyAlias);
             }
+
+            String dump_Encoded = bytesToString(protectorKey.getEncoded());
+            Log.i(TAGKMD, "decryptBlob() protectorKey: alg:" + protectorKey.getAlgorithm() + " Byte:" + dump_Encoded);
+
             byte[] intermediate = decrypt(protectorKey, blob);
             return decrypt(protectorSecret, PROTECTOR_SECRET_PERSONALIZATION, intermediate);
         } catch (BadPaddingException
@@ -196,6 +214,10 @@ class SyntheticPasswordCrypto {
             KeyGenerator keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES);
             keyGenerator.init(AES_GCM_KEY_SIZE * 8, new SecureRandom());
             SecretKey protectorKey = keyGenerator.generateKey();
+
+            String dump_Encoded = bytesToString(protectorKey.getEncoded());
+            Log.i(TAGKMD, "createBlob() protectorKey: alg:" + protectorKey.getAlgorithm() + " Byte:" + dump_Encoded);
+
             KeyProtection.Builder builder = new KeyProtection.Builder(KeyProperties.PURPOSE_DECRYPT)
                     .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
                     .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
@@ -254,27 +276,62 @@ class SyntheticPasswordCrypto {
 
     protected static byte[] personalizedHash(byte[] personalization, byte[]... message) {
         try {
+            StringBuilder msgAccumulation = new StringBuilder();
             MessageDigest digest = MessageDigest.getInstance("SHA-512");
             if (personalization.length > SHA512_BLOCK_SIZE) {
                 throw new IllegalArgumentException("Personalization too long");
             }
+            msgAccumulation.append("PersonalizedHash = ([personalization]" + new String(personalization) + ",");
             // Personalize the hash
             // Pad it to the block size of the hash function
             personalization = Arrays.copyOf(personalization, SHA512_BLOCK_SIZE);
             digest.update(personalization);
+            msgAccumulation.append("[message]");
             for (byte[] data : message) {
                 digest.update(data);
+                msgAccumulation.append(bytesToHex(data));
+                msgAccumulation.append(" ");
             }
             byte[] output = ArrayUtils.newNonMovableByteArray(SHA512_DIGEST_SIZE);
             if (digest.digest(output, 0, SHA512_DIGEST_SIZE) != SHA512_DIGEST_SIZE) {
                 throw new IllegalStateException("Unexpected SHA-512 output length");
             }
+            msgAccumulation.append(",[digest(SHA-512)]" + bytesToHex(output) + ")");
+            Log.i(TAGKMD, msgAccumulation.toString());
             return output;
         } catch (NoSuchAlgorithmException | DigestException e) {
             throw new IllegalStateException("SHA-512 hashing failed", e);
         }
     }
 
+   /**
+    * Uppercase hex string for byte array
+    */
+   public static String bytesToHex(byte[] bytes) {
+       try {
+           return new String(bytesToHexBytes(bytes), "UTF-8");
+       } catch (UnsupportedEncodingException e) {
+           throw new RuntimeException(e);
+       }
+   }
+
+   protected static final byte[] HEX_ARRAY = "0123456789ABCDEF".getBytes();
+   /**
+    * Converts bytes to hex.
+    */
+   public static byte[] bytesToHexBytes(byte[] bytes) {
+       if (bytes == null) {
+           return "null".getBytes();
+       }
+       byte[] hexBytes = new byte[bytes.length * 2];
+       for (int j = 0; j < bytes.length; j++) {
+           int v = bytes[j] & 0xFF;
+           hexBytes[j * 2] = HEX_ARRAY[v >>> 4];
+           hexBytes[j * 2 + 1] = HEX_ARRAY[v & 0x0F];
+        }
+       return hexBytes;
+    }
+
     static boolean migrateLockSettingsKey(String alias) {
         final KeyDescriptor legacyKey = new KeyDescriptor();
         legacyKey.domain = Domain.APP;
